jaxtronomy/LensModel/Solver/lens_equation_solver.py

```diff
@@ -0,0 +1,773 @@
+import copy
+
+import jax.lax as lax
+
+import jax.random as random
+import jax.numpy as jnp
+from jax import jit
+import numpy as np
+import jaxtronomy.Util.util as util
+import lenstronomy.Util.image_util as image_util
+from jax.scipy.optimize import minimize
+from lenstronomy.LensModel.Solver.epl_shear_solver import solve_lenseq_pemd
+
+__all__ = ["LensEquationSolver"]
+
+SUPPORTED_LENS_MODELS_ANALYTICAL = (
+    ["SIE", "SHEAR"],
+    ["SIE"],
+    ["EPL_NUMBA", "SHEAR"],
+    ["EPL_NUMBA"],
+    ["EPL", "SHEAR"],
+    ["EPL"],
+)
+
+
+class LensEquationSolver(object):
+    """Class to solve for image positions given lens model and source position."""
+
+    def __init__(self, lensModel):
+        """This class must contain the following definitions (with same syntax as the
+        standard LensModel() class: def ray_shooting() def hessian() def magnification()
+
+        :param lensModel: instance of a class according to
+            lenstronomy.LensModel.lens_model
+        """
+        self.lensModel = lensModel
+
+    # --------------------------------------------------------------------------------
+    # The following two methods are required to allow the JAX compiler to recognize
+    # the LensEquationSolver class. Methods involving the self variable can be
+    # jit-decorated, and will recompile whenever the self variables listed under
+    # aux_data are changed
+    def _tree_flatten(self):
+        children = ()
+        aux_data = {"lensModel": self.lensModel}
+        return (children, aux_data)
+
+    @classmethod
+    def _tree_unflatten(cls, aux_data, children):
+        return cls(*children, **aux_data)
+
+    # --------------------------------------------------------------------------------
+
+    def image_position_from_source(
+        self, sourcePos_x, sourcePos_y, kwargs_lens, solver="lenstronomy", **kwargs
+    ):
+        """Solves the lens equation, i.e. finds the image positions in the lens plane
+        that are mapped to a given source position.
+
+        :param sourcePos_x: source position in units of angle
+        :param sourcePos_y: source position in units of angle
+        :param kwargs_lens: lens model parameters as keyword arguments
+        :param solver: which solver to use, can be 'lenstronomy' (default), 'analytical'
+            or 'stochastic'.
+        :param kwargs: Any additional kwargs are passed to the chosen solver, see the
+            documentation of image_position_lenstronomy, image_position_analytical and
+            image_position_stochastic
+        :returns: (exact) angular position of (multiple) images ra_pos, dec_pos in units
+            of angle
+        """
+        if solver == "lenstronomy":
```

Comment:
do these IF clauses work in JAX? Or do we need a different class hierarchy such that when calling the definition, no if/else statements are needed (i.e. define it on the class level instead)?

Reply:
Same as my comment [above](https://github.com/lenstronomy/JAXtronomy/pull/14#discussion_r1694451063)

Reply:
The main struggle with jaxifying some of these if-statements is that when using jax.lax.cond, whether the condition is true or false, the resulting output array must have the same shape, which doesn't work when you're appending or removing elements of an array. For example, here's the error you would get if you were to try to jaxify the `if magnification_limit is Not None`: `true_fun and false_fun output must have same type structure, got PyTreeDef(([*, *], [*, *], [*, *])) and PyTreeDef(([*], [*], [*]))` since if magnification_limit is None, you return the whole array, but if magnification_limit is not None, you remove some elements (jnp.where has a similar issue)

Reply:
@ahuang314 I think we should re-design then the solver class a bit and request these options to be put into the class instance and once the class is initialized, that there are little to no ambiguities left. It's ok to deviate a bit from what lenstronomy has. In turn, we can also change a bit the structure in lenstronomy to be better compatible with the jax version

Reply:
@sibirrer Hmm, If I'm understanding this correctly, you're suggesting that instead of having `magnification_limit` be a function argument, we instead make it a static class variable that is defined when an instance of the class is created. That way, the `if statement` is not needed and we can go through with downselecting the solutions.

This will still run into the same fundamental problem: JAX requires that all array shapes and sizes are known at compile time. This means that functions that branch off of conditions must have the same return structures, and repeated executions of the same function must have the same return structure regardless of the function arguments. The problem with jaxifying the lens equation solver completely is that you don't know how many solutions there will be until after you've solved it, and you also don't know how many solutions are downselected from the magnification limit until you do it. This is why I did not jaxify the `image_position_lenstronomy` and `candidate_solutions` functions in the solver. Here is a screenshot showing the error you would get if you tried the downselecting:
![image](https://github.com/user-attachments/assets/4344ec4a-df01-42c4-bcdf-c3f56b16c32b)
![image](https://github.com/user-attachments/assets/36e88dab-a8e3-4e6c-bbf6-3cea2c63510f)

The way it is done in herculens is, the user specifies the number of solutions `nsolutions` that the solver expects to find. This is treated as a static variable so the solver will always return exactly an array of size `nsolutions`. However, I dislike this method of doing it because if the solver actually finds more solutions than that, some will be excluded, and if the solver finds less than that, one of the solutions will be duplicated until the array is of size `nsolutions`. The magnification limit downselection in herculens is also done in an unjaxified way.

To conclude, I think we should only focus on jaxifying only the heavy-computation subfunctions and not worry about large scale restructuring in an attempt to jaxify everything. The following three line profiles show that, even if we were to fully jaxify the entirety of `image_position_lenstronomy` and `candidate_solutions`, the benefit we would get is extremely tiny due to how little time is actually spent on the currently unjaxified part of the code.
[Line profile for unjaxified solver with jaxified ray shooting](https://pastebin.com/MPShhPen)
[Line profile for solver after jaxifying gradient descent with jaxified ray shooting](https://pastebin.com/ZRRVDGmw)
[Line profile for `candidate_solutions`](https://pastebin.com/fQbeRfTF)

Reply:
I see. It's very tricky to jaxify all the parts even when class structure change, as the issues are algorithmically.
The only advantage of jaxifying everything would be an auto-differentiation. But since we barely use the lens equation solver in the actual fitting, we do not have to solve a problem we do not have for now.